server/src/main/java/org/apache/iotdb/db/qp/strategy/optimizer/ConcatPathOptimizer.java

```diff
@@ -238,15 +239,101 @@ private FilterOperator concatFilter(List<Path> fromPaths, FilterOperator operato
       // Note that,
       // two fork tree has to be maintained while removing stars in paths for DnfFilterOptimizer
       // requirement.
-      return constructTwoForkFilterTreeWithAnd(noStarPaths, operator);
+      return constructBinaryFilterTreeWithAnd(noStarPaths, operator);
+    }
+  }
+
+  // e.g. translate "select * from root.ln.d1, root.ln.d2 where s1 < 20 AND s2 > 10"
+  // to "select * from root.ln.d1, root.ln.d2 where
+  // (root.ln.d1.s1 < 20 AND root.ln.d1.s2 > 10) OR (root.ln.d2.s1 < 20 AND root.ln.d2.s2 > 10)"
+  private FilterOperator concatFilterByDivice(List<Path> fromPaths, FilterOperator operator)
+          throws LogicalOptimizeException {
+    // remove stars in fromPaths and get deviceId
+    List<String> noStarDevices = removeStarsInDeviceWithUnique(fromPaths);
+
+    FilterOperator filterBinaryTree = new FilterOperator(SQLConstant.KW_OR);
+    FilterOperator currentNode = filterBinaryTree;
+    FilterOperator parentNode = currentNode;
+    // to check whether duplicate
+    // e.g. SELECT * FROM root.ln.d1, root.ln.d2 where time < 10
+    // brings two 'time < 10' filter operator
+    Set<FilterOperator> operatorSet = new HashSet<>();
+    for (int i = 0; i < noStarDevices.size(); i++) {
+      FilterOperator newOperator = operator.clone();
```

Comment:
do we need clone?

Reply:
Yes. Otherwise the next loop the operator will change.